File: Blind_75_LeetCode_Questions/c++/Matrix/Q54_Spiral_Matrix/Spiral_Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) 
    {
        int rows = matrix.size();
        int cols = matrix[0].size();
        vector<int> directions = {0, 1, 0, -1, 0};
        vector<int> result;
        vector<vector<bool>> visited(rows, vector<bool>(cols, false));

        int row = 0, col = 0, dirIndex = 0;
        
        for(int remain = rows * cols; remain > 0; --remain)
        {
            result.push_back(matrix[row][col]);
            visited[row][col] = true;

            int nextRow = row + directions[dirIndex];
            int nextCol = col + directions[dirIndex + 1];

            if((nextRow < 0) || (nextRow >= rows) || (nextCol < 0) || (nextCol >= cols) || (visited[nextRow][nextCol]))
            {
                dirIndex = (dirIndex + 1) % 4;
            }

            row = row + directions[dirIndex];
            col = col + directions[dirIndex + 1];
        }
        return result;
        
    }
};

File: Blind_75_LeetCode_Questions/java/Matrix/Q54_Spiral_Matrix/Spiral_Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) 
    {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int [] dirRow = {0, 1, 0, -1};
        int [] dirCol = {1, 0, -1, 0};
        int row = 0, col = 0, dirIndex = 0;
        List<Integer> result = new ArrayList<>();
        boolean[][] visited = new boolean[rows][cols];
        
        for(int h = rows * cols; h > 0; --h)
        {
            result.add(matrix[row][col]);
            visited[row][col] = true;
            int nextRow = row + dirRow[dirIndex];
            int nextCol = col + dirCol[dirIndex];
            if(nextRow < 0 || nextRow >= rows || nextCol < 0 || nextCol >= cols || visited[nextRow][nextCol])
            {
                dirIndex = (dirIndex + 1) % 4;
                nextRow = row + dirRow[dirIndex];
                nextCol = col + dirCol[dirIndex];
            }
            row = nextRow;
            col = nextCol;
        }
        return result;


    }
}
